### PR TITLE
Add face recognition example using center loss and lfw dataset

### PR DESCRIPTION
**Pull request by louis-she**
please see [readme.md](https://github.com/louis-she/examples/blob/da787cf53dcb4a2b7427b67252712bd9d23bf0ca/face_recognition/README.md)

### CODE REVIEW

**Comment by sammiie** [comment]
you can create a face recognition example using Python, a deep learning framework (e.g., PyTorch), and the LFW (Labeled Faces in the Wild) dataset. Below is a step-by-step guide with code snippets to help you add this example to your project:

1:Import Required Libraries:
Start by importing the necessary libraries, including PyTorch, torchvision, and other relevant modules.
python
import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms
import torchvision.datasets as datasets
import torch.optim as optim

Load and Preprocess the LFW Dataset:
Download and preprocess the LFW dataset. You may need to install the lfw-dataset package to access the dataset.
# Assuming you have installed the lfw-dataset package
from lfw_dataset import LFWLoader

lfw_loader = LFWLoader(
    data_path="path/to/lfw/dataset",
    image_size=(112, 112),
    batch_size=64,
    num_workers=4,
)

Define the Center Loss:
Define the center loss, which is a loss function used in face recognition tasks to improve the compactness of learned features.
class CenterLoss(nn.Module):
    def __init__(self, num_classes, feat_dim, device):
        super(CenterLoss, self).__init__()
        self.num_classes = num_classes
        self.feat_dim = feat_dim
        self.device = device
        self.centers = nn.Parameter(torch.randn(num_classes, feat_dim).to(device))

    def forward(self, x, labels):
        center_exp = self.centers.index_select(0, labels.long())
        count = torch.histc(labels.to(torch.float32), bins=self.num_classes, min=0, max=self.num_classes - 1)
        count_exp = count.index_select(0, labels.long())
        return torch.sum(torch.norm(x - center_exp, 2, 1) / count_exp)

Define a CNN Model:
Define a convolutional neural network (CNN) model for face recognition. You can use existing architectures like ResNet or create a custom model.
class FaceRecognitionModel(nn.Module):
    def __init__(self, num_classes):
        super(FaceRecognitionModel, self).__init__()
        # Define your model architecture here
        self.conv1 = nn.Conv2d(in_channels=3, out_channels=64, kernel_size=3, padding=1)
        # ...

    def forward(self, x):
        # Implement the forward pass of your model
        # ...

Train the Model with Center Loss:
Implement the training loop, including the use of the Center Loss in addition to a classification loss (e.g., cross-entropy).
# Define your model and optimizer
model = FaceRecognitionModel(num_classes=lfw_loader.num_classes)
optimizer = optim.Adam(model.parameters(), lr=0.001)
center_loss = CenterLoss(num_classes=lfw_loader.num_classes, feat_dim=256, device="cuda")

# Training loop
for epoch in range(num_epochs):
    for images, labels in lfw_loader.train_loader:
        images, labels = images.to("cuda"), labels.to("cuda")

        # Forward pass
        features = model(images)
        clf_loss = nn.CrossEntropyLoss()(features, labels)
        ctr_loss = center_loss(features, labels)
        total_loss = clf_loss + ctr_loss

        # Backpropagation and optimization
        optimizer.zero_grad()
        total_loss.backward()
        optimizer.step()

# Evaluation code here
# ...

Evaluate the Model:
Implement code to evaluate the trained model on the LFW test dataset, calculate accuracy, and perform face recognition.

Additional Steps:

You may need to fine-tune hyperparameters, adjust the model architecture, and optimize the training process for your specific use case.
Implement a face recognition pipeline, including face detection and alignment, before feeding images into your model.
This code serves as a starting point for adding a face recognition example using center loss and the LFW dataset. Be sure to adapt and extend it based on your project's requirements and objectives.
